main.tf

```diff
@@ -0,0 +1,119 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+module "resource_names" {
+  source  = "d2lqlh14iel5k2.cloudfront.net/module_library/resource_name/launch"
+  version = "~> 1.0"
+
+  for_each = var.resource_names_map
+
+  region                  = join("", split("-", var.location))
+  class_env               = var.environment
+  cloud_resource_type     = each.value.name
+  instance_env            = var.environment_number
+  instance_resource       = var.resource_number
+  maximum_length          = each.value.max_length
+  logical_product_family  = var.logical_product_family
+  logical_product_service = var.logical_product_service
+  use_azure_region_abbr   = var.use_azure_region_abbr
+}
+
+module "resource_group" {
+  source  = "d2lqlh14iel5k2.cloudfront.net/module_primitive/resource_group/azurerm"
+  version = "~> 1.0"
+
+  name     = local.resource_group_name
+  location = var.location
+  tags     = local.resource_group_tags
```

Comment:
According to the guidelines we follow from Matt, for each resource we should always have a tag named resource_name that should point to the standard name coming from the naming_resource module as shown here https://github.com/launchbynttdata/tf-azurerm-module_reference-kubernetes_cluster/blob/main/main.tf#L38

Reply:
Should we consider having the primitive modules add their own tag with their resource name? This would allow us to use arbitrary naming strategies in the collection and reference layers and whatever name variable we pass to those primitives would be auto-added to their tags.

Reply:
The resource_name uses the standard name generated by the naming module but our primitive modules are not integrated with naming module (by design) so that the collection or ref arch modules could name the primitive resources the way they want. Thats the reason, the primitive modules cannot assign the resource_name tag

Reply:
I'm not suggesting we integrate the resource_name module to our primitives, quite the opposite.

In a collection or reference module, we know that we need to pull in resource_name to generate the correct naming patterns for each of the resources we're trying to spin up. What I'm suggesting is that in a hypothetical child module's view we do something like:

```terraform
# tf-azurerm-module_primitive-example
locals {
  default_tags = {
    resource_name = var.name # <--- the only thing we'd need to add to what we already should be doing
    provisioner = "Terraform"
  }

  tags = merge(local.default_tags, var.tags)
}

resource "cloud_provider" "example_resource" {
  name = var.name
  ...
  tags = local.tags
}

```

This way, whatever name you pass to the resource automatically gets added as a resource_name tag. If you passed the full "launch-dso-eastus-dev-000-whatever-000" name, it would put that in the tag. If you passed a random suffix or even just "hello world" as the resource's name, that would end up in the tag.

This saves us from having to generate additional tags for each of the downstream components at the collection/reference level and instead delegates that naming tag to the primitive module. It would still be possible to pass additional tags to the primitive, we just expand the default behavior to auto-tag that resource_name key with whatever name is passed to the TF module rather than building the tag in the higher-order module and passing it down separately.